File: compiler/src/lexer.cc
#include "lexer.h"

namespace Lexer {
bool is_l(char c) { return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z'); }
bool is_w(char c) { return c == ' ' || c == '\n' || c == '\t'; }
bool is_n(char c, TokenState s) {
  return (c == '.' || c == '_') ? s == READ_NUM : (c >= '0' && c <= '9');
}
bool is_o(char c) {
  return operators.find(std::string(1, c)) != operators.end();
}

std::string pop_buf(std::vector<char> &buf) {
  auto y = std::string(buf.begin(), buf.end());
  buf.clear();
  return y;
}

TokenState grab_state(char x, TokenState s) {
  if (is_l(x)) {
    return READ_WORD;
  } else if (is_n(x, s)) {
    return READ_NUM;
  } else if (x == '\"') {
    return READ_STRING;
  } else if (x == '\'') {
    return READ_CHAR;
  } else if (is_o(x)) {
    return READ_OP;
  }
  return START;
}

std::vector<Token> tokenize(std::string input) {
  std::vector<Token> t = {};
  std::vector<char> b = {};
  TokenState s = START;
  for (auto x : input) {
    switch (s) {
    case START:
      if (!is_w(x)) {
        b.push_back(x);
      }
      s = grab_state(x, s);
      break;
    case READ_WORD:
      if (!(is_l(x) || is_n(x, s))) {
        auto y = pop_buf(b);
        t.push_back({
            .t = kwds.find(y) != kwds.end() ? kwds.at(y) : IDENTIFIER,
            .v = kwds.find(y) != kwds.end() ? "" : y,
        });
        s = grab_state(x, START);
      }
      if (!is_w(x)) {
        b.push_back(x);
      }
      break;
    case READ_NUM:
      if (!is_n(x, s)) {
        auto y = pop_buf(b);
        t.push_back({
            .t = y.find('.') != std::string::npos ? FLOAT : INT,
            .v = y,
        });
        s = grab_state(x, START);
      }
      if (!is_w(x)) {
        b.push_back(x);
      }
      break;
    case READ_OP:
      if (!is_o(x)) {
        if (b.size() == 1 && b.at(0) == '.' && is_n(x, s)) {
          s = grab_state(x, START);
        } else {
          auto y = pop_buf(b);
          t.push_back({.t = operators.at(y), .v = ""});
          s = grab_state(x, START);
        }
      } else {
        auto z = std::string(b.begin(), b.end()).append(std::string(1, x));
        if (operators.find(z) == operators.end()) {
          auto y = pop_buf(b);
          t.push_back({.t = operators.at(y), .v = ""});
        }
      }
      if (!is_w(x)) {
        b.push_back(x);
      }
      break;
    case READ_STRING:
    case READ_CHAR:
    default:
      break;
    }
  }
  auto y = pop_buf(b);
  switch (s) {
  case READ_WORD:
    t.push_back({
        .t = kwds.find(y) != kwds.end() ? kwds.at(y) : IDENTIFIER,
        .v = kwds.find(y) != kwds.end() ? "" : y,
    });
    break;
  case READ_NUM:
    t.push_back({
        .t = y.find('.') != std::string::npos ? FLOAT : INT,
        .v = y,
    });
    break;
  case READ_OP:
    t.push_back({.t = operators.at(y), .v = ""});
    break;
  default:
    break;
  }
  t.push_back({.t = END_OF_FILE, .v = ""});
  return t;
}
}

File: compiler_v2/src/lexer.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

const Token = struct {
    t: u8,
    v: []const u8,
};

const TokenList = std.ArrayList(Token);

const OPS: [][]const u8 = .{
    ";",  ":",  ".",  ",",  "(",  "[",  "{",  ")",  "]",  "}",  "=",
    "!",  "<",  ">",  "*",  "+",  "/",  "-",  "&",  "|",  "==", "!=",
    "<=", ">=", "*=", "+=", "/=", "-=", "&=", "&&", "|=", "||", "=>",
};

const ParserState = enum {
    START,
    READ_WORD,
    READ_NUM,
    READ_STRING,
    READ_CHAR,
    READ_OP,
};

fn is_c(x: u8) bool {
    return (x >= 'a' and x <= 'z') or (x >= 'A' and x <= 'Z');
}
fn is_n(x: u8, s: ParserState) bool {
    if (x == '.') {
        return s == .READ_NUM;
    }
    return x >= '0' and x <= '9';
}
fn is_o(x: []const u8) bool {
    for (0..OPS.len) |i| {
        if (OPS[i] == x) {
            return true;
        }
    }
    return false;
}
fn is_w(x: u8) bool {
    return x == '\r' or x == '\n' or x == '\t' or x == ' ';
}
fn grab_state(x: u8, s: ParserState) ParserState {
    if (x == '\"') {
        return .READ_STRING;
    }
    if (is_c(x)) {
        return .READ_WORD;
    }
    if (is_n(x, s)) {
        return .READ_NUM;
    }
    if (is_o(.{x})) {
        return .READ_OP;
    }
    return .START;
}
fn clear_buf(alloc: Allocator, x: *std.ArrayList(u8)) []const u8 {
    const y = try std.mem.join(alloc, "", x);
    x.clearAndFree();
    return y;
}
pub fn lex(alloc: std.mem.Allocator, input: []const u8) std.ArrayList(Token) {
    var s: ParserState = .START;
    var t = std.ArrayList(Token).init(alloc);
    var b = std.ArrayList(u8).init(alloc);
    for (0..input.len) |i| {
        switch (s) {
            .START => {
                s = grab_state(input[i], s);
                if (s == .START and !is_w(input[i])) {
                    const x: Token = .{ .t = 1, .v = "" };
                    try t.append(x);
                }
                break;
            },
            .READ_WORD => {
                if (!(is_c(input[i]) or is_n(input[i], s))) {
                    try t.append(.{ .t = 2, .v = clear_buf(&b) });
                    s = grab_state(input[i], .START);
                }
                break;
            },
            .READ_NUM => {
                if (!is_n(input[i], s)) {
                    try t.append(.{ .t = 3, .v = clear_buf(&b) });
                    s = grab_state(input[i], .START);
                    b.clearAndFree();
                }
                break;
            },
            .READ_STRING => {
                if (input[i] == '\"') {
                    b.append(input[i]);
                    try t.append(.{ .t = 4, .v = clear_buf(&b) });
                    s = .START;
                }
                break;
            },
            .READ_CHAR => {
                break; // this needs done
            },
            .READ_OP => {
                const c = b ++ input[i];
                if (!is_o(c)) {
                    try t.append(.{ .t = 6, .v = clear_buf(&b) });
                    s = grab_state(input[i], .START);
                }
                break;
            },
        }
        if (s != .START) {
            b.append(input[i]);
        }
    }
    try t.append(.{ .t = 0, .v = clear_buf(alloc, b) });
    return t;
}
